### PR TITLE
refactor(db): consolidate health telemetry into health_metric_samples & fix weight unit conversions

### PR DESCRIPTION
**Pull request by CodeWithCJ**
> [!TIP]
> **Help us review and merge your PR faster!**
> Please ensure you have completed the **Checklist** below.
> For **Frontend** changes, please run `pnpm run validate` to check for any errors.
> PRs that include tests and clear screenshots are highly preferred!
> Note: AI-generated descriptions must be manually edited for conciseness. Do not paste raw AI summaries.

## Description

**What problem does this PR solve?**
(Keep it concise. 1–2 sentences.)

## Summary
This PR consolidates multiple separate time-series health metric tables into a unified `health_metric_samples` table, refactors backend/frontend consumers to query the consolidated schema, and fixes Garmin strength training set weight unit conversions.

**How did you implement the solution?**
(Brief technical approach.)

---
### 1. Database & Schema Consolidation
- **Unified Health Samples Table**: Consolidated `heart_rate_entries`, `hrv_entries`, `respiration_entries`, and `spo2_entries` into a single `health_metric_samples` table indexed by `metric_type`.
- **Zod Schemas**: Created `HealthMetricSamples.zod.ts` in `@workspace/shared` and removed deprecated per-metric Zod schemas (`HeartRateEntries`, `HrvEntries`, `RespirationEntries`, `Spo2Entries`).
- **RLS & Documentation**: Updated Row-Level Security policies and developer docs (`docs/content/8.developer/4.database.md` and `11.database-security-tiers.md`).
### 2. Backend & Data Ingestion (`SparkyFitnessServer`)
- **Garmin Health Ingestion**: Refactored `garminHealthProcessor.ts` to batch write time-series metrics into `health_metric_samples`.
- **Repositories & Routes**: Updated `genericHealthRepository`, `workoutTelemetryRepository`, `syncedDataRepository`, and `genericHealthRoutes` to use the unified table queries.
### 3. Frontend & Visualizations (`SparkyFitnessFrontend`)
- Updated `useGenericHealth` hook and generic health services.
- **Unit Conversion Fixes**: Fixed Garmin strength set weight conversions (lbs ↔ kg) in `WorkoutReportVisualizer` and `WorkoutSessionBreakdown`.

Linked Issue: Closes #

## How to Test

1. Check out this branch and run `...`
2. Navigate to...
3. Verify that...

## PR Type

- [x] Issue (bug fix)
- [x] New Feature
- [ ] Refactor
- [ ] Documentation

## Checklist

**All PRs:**

- [x] **[MANDATORY - ALL] Integrity & License**: I certify this is my own work, free of malicious code, and I agree to the [License terms](https://github.com/CodeWithCJ/SparkyFitness/blob/main/LICENSE).

**New features only:**

- [x] **[MANDATORY for new feature] Alignment**: I have raised a GitHub issue and it was reviewed/approved by maintainers or it was approved on Discord.

**Frontend changes (`SparkyFitnessFrontend/`):**

- [x] **[MANDATORY for Frontend changes] Quality**: I have run `pnpm run validate` and it passes.
- [x] **[MANDATORY for Frontend changes] Translations**: I have only updated the English (`en`) translation file.

**Backend changes (`SparkyFitnessServer/`):**

- [x] **[MANDATORY for Backend changes] Code Quality**: I have run typecheck, lint, and tests. New files use TypeScript, new endpoints have Zod schemas, and new endpoints include tests.
- [x] **[MANDATORY for Backend changes] Database Security**: I have updated `rls_policies.sql` for any new user-specific tables.

**UI changes (components, screens, pages):**

- [x] **[MANDATORY for UI changes] Screenshots**: I have attached Before/After screenshots below.

**Mobile changes (`SparkyFitnessMobile/`):**

- [ ] **[MANDATORY for Mobile changes] Tested on device or emulator**: I have verified the changes work on iOS or Android.

## Screenshots

<details>
<summary>Click to expand</summary>

### Before

before

### After

after

</details>
image
image
image
image
image
image
image
image
image
image
image
image

## Notes for Reviewers

> Optional — use this for anything that doesn't fit above: known tradeoffs, areas you'd like specific feedback on, qustions you have or context that helps reviewers.

## Summary by CodeRabbit

* **New Features**
  * Added workout analytics for cardio trends, personal records, matched courses, activity search, laps, GPS routes, heart-rate zones, and session breakdowns.
  * Added wearable health summaries covering daily wellness, vitals, stress, body battery, VO2 Max, recovery, and readiness.
  * Added smart-scale tracking for muscle mass, bone mass, and body water.
  * Expanded Garmin and Strava activity telemetry and workout detail views.
* **Bug Fixes**
  * Improved unit conversion, activity identification, missing-data handling, and provider detail editing.
* **Documentation**
  * Updated health-data permissions, deletion, and check-in field documentation.

### CODE REVIEW

**Comment by CodeWithCJ** [comment]
@coderabbitai review again

File: docs/content/8.developer/5.api-reference.md [hunk 1]
```diff
@@ -219,22 +219,22 @@ Example Error Response:
...
+          "bone_mass_kg": <number, optional>,
+          "body_water_percentage": <number, optional>
         }
         ```
     *   **Fields**:
         *   `entry_date` (required, string): The date of the check-in in `YYYY-MM-DD` format.
-        *   Other fields (optional, number): Various measurement values.
+        *   Other fields (optional, number): Measurement values. Masses are in kilograms, circumferences and height in centimetres, and `body_fat_percentage` / `body_water_percentage` are percentages.
+        *   BMI is not accepted or stored — it is derived from weight and height where it is displayed.
+        *   Sending `null` for a field clears a previously recorded value for that day.
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply docs/content/8.developer/5.api-reference.md:237]: Fixed in 282fdf75. GET/PUT response examples now use `body_fat_percentage` and include `muscle_mass_kg`, `bone_mass_kg`, `body_water_percentage`; the PUT field list mentions the new fields.

File: SparkyFitnessServer/ai/tools/checkinTools.ts [hunk 2]
```diff
@@ -275,6 +275,26 @@ Actions:
...
+              if (isSet(args.bone_mass)) {
+                measurements.bone_mass_kg = convertWeight(
+                  args.bone_mass,
+                  wUnit,
+                  'kg'
+                );
+              }
+              if (isSet(args.body_water)) {
+                measurements.body_water_percentage = args.body_water;
+              }
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/ai/tools/checkinTools.ts:297]: Fixed in 282fdf75. The confirmation text now uses the resolved `wUnit`/`hUnit`/`mUnit` rather than literal `kg`/`cm` fallbacks — the same defect affected `weight`, `height`, and the circumference fields, so all of them were corrected. An existing test pinned the old behaviour ("text still defaults to kg"); it now asserts the unit actually used for conversion.

File: SparkyFitnessServer/db/migrations/20260730000000_create_generic_health_and_workout_tables.sql [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+DO $$
+BEGIN
+  IF NOT EXISTS (
+    SELECT 1 FROM pg_constraint WHERE conname = 'uq_vitals_entry'
+  ) THEN
+    ALTER TABLE vitals_entries
+      ADD CONSTRAINT uq_vitals_entry UNIQUE (user_id, source_provider, "timestamp");
+  END IF;
+END $$;
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/db/migrations/20260730000000_create_generic_health_and_workout_tables.sql:278]: Fixed in 282fdf75. Added `AND conrelid = 'public.vitals_entries'::regclass`, matching the existing convention in `20251023221501_add_meal_id_to_food_entries.sql`.

File: SparkyFitnessServer/db/migrations/20260730000000_create_generic_health_and_workout_tables.sql [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    SELECT *
+    FROM (VALUES
+      -- category name,             target column,            max plausible value
+      ('Muscle Mass',           'muscle_mass_kg',        1000::numeric),
+      ('Bone Mass',             'bone_mass_kg',          1000::numeric),
+      -- Garmin-only. Withings reports water as MASS in kg ('Hydration', and
+      -- 'Body Water Breakdown' for extra/intracellular), never a percentage,
+      -- so those categories are deliberately out of scope here.
+      ('Body Water Percentage', 'body_water_percentage',  100::numeric)
+    ) AS t(category_name, column_name, max_value)
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/db/migrations/20260730000000_create_generic_health_and_workout_tables.sql]: Fixed in 282fdf75, though via the ceiling rather than the comparison: the mass limits are now `999.99` (the largest value `numeric(5,2)` holds) and the comparison stays `<=`. A strict `<` would also have excluded a legitimate `body_water_percentage` of exactly 100, which stores fine.

File: SparkyFitnessServer/schemas/measurementSchemas.ts [hunk 5]
```diff
@@ -86,6 +86,13 @@ export const UpsertCheckInBodySchema = z
     steps: nullableOptionalLegacyNumber,
     height: nullableOptionalLegacyNumber,
     body_fat_percentage: nullableOptionalLegacyNumber,
+    // Smart-scale composition. These must be declared even though the schema is
+    // .loose(): without them the values still reach the repository but skip
+    // coercion entirely, so a non-numeric or oversized value surfaced as a 500
+    // from the numeric(5,2) column instead of a 400.
+    muscle_mass_kg: nullableOptionalLegacyNumber,
+    bone_mass_kg: nullableOptionalLegacyNumber,
+    body_water_percentage: nullableOptionalLegacyNumber,
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/schemas/measurementSchemas.ts]: Fixed in 282fdf75. Added a bounded coercion helper and applied it to all three fields on both `UpsertCheckInBodySchema` and `UpdateCheckInBodySchema` — masses `0..999.99` (the `numeric(5,2)` ceiling), body water `0..100`. Covered in `tests/measurementSchemas.checkIn.test.ts`, including that the boundary values still pass.

File: SparkyFitnessServer/services/exerciseStatsService.ts [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+
+    const breakdownSql = `
+      SELECT 
+        DATE_TRUNC('${truncUnit}', entry_date) as period_start,
+        COALESCE(SUM(COALESCE(distance, 0)), 0) as distance_km,
+        COALESCE(SUM(duration_minutes), 0) as duration_minutes,
+        COALESCE(SUM(calories_burned), 0) as calories_burned,
+        COUNT(DISTINCT id) as workout_count,
+        AVG(avg_heart_rate) as avg_heart_rate,
+        COALESCE(SUM(elevation_gain_meters), 0) as elevation_gain_meters
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/services/exerciseStatsService.ts:219]: Fixed in 282fdf75. `totalSql` now sums `elevation_gain_meters` and the result feeds `totals.totalElevationGainMeters`, with an assertion added to `tests/exerciseStatsService.test.ts`.

File: SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+export function extractGarminGpsPoints(
+  payload: UnknownRecord
+): ExtractedGpsPoint[] {
+  if (Array.isArray(payload.gps_points)) {
+    return asArray<UnknownRecord>(payload.gps_points).map(
+      (pt: UnknownRecord) => {
+        const tsVal = pt.timestamp;
+        const timestamp = tsVal
+          ? new Date(tsVal as string | number | Date)
+          : new Date();
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts]: Fixed in 282fdf75. Added a `toTrackpointTimestamp()` helper and switched all three branches (`gps_points`, `activityDetailMetrics`, `geoPolylineDTO.polyline`) to `flatMap`, dropping points with a missing or unparseable time — the same rule `extractGarminLaps` already applied. Covered in `tests/telemetryExtractors.laps.test.ts`.

File: SparkyFitnessServer/tests/fitImportService.test.ts [hunk 8]
```diff
@@ -15,8 +15,19 @@ vi.mock('../db/poolManager', () => ({
...
   },
 }));
+vi.mock('../models/workoutTelemetryRepository', () => ({
+  _bulkInsertExerciseEntryLapsWithClient: vi.fn(),
+  _bulkInsertExerciseEntryGpsPointsWithClient: vi.fn(),
+  _bulkInsertExerciseEntryHrZonesWithClient: vi.fn(),
+  bulkInsertExerciseEntryLaps: vi.fn(),
+  bulkInsertExerciseEntryGpsPoints: vi.fn(),
+  bulkInsertExerciseEntryHrZones: vi.fn(),
+}));
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/tests/fitImportService.test.ts]: Addressed in 282fdf75, with a correction to the premise: this test passes as-is (7/7), and both specifier styles are used throughout the existing suite, so Vitest resolves them either way. The `.js` specifiers are added for consistency. The more useful find here was the stale comment claiming `workoutTelemetryRepository` "isn't mocked at module level" while it is mocked directly above — that has been corrected.

File: SparkyFitnessServer/ai/tools/devTools.ts [hunk 9]
```diff
@@ -26,6 +58,40 @@ async function assertDevAccess(userId: string): Promise<string | null> {
...
+    } catch {
+      // Connection already unusable; releasing below is all we can do.
+    }
+    throw error;
+  } finally {
+    if (client && typeof client.release === 'function') {
+      client.release();
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/ai/tools/devTools.ts:93]: Agreed on the analysis, deferring the fix — tracked as #2001.

`BEGIN READ ONLY` blocks writes but places no restriction on reads, and `getSystemClient()` is the owner role, so an arbitrary `SELECT` could still reach `pg_read_file()` or other users' rows. Since access requires both `DEV_TOOLS_ENABLED=true` and an admin role, this is a defence-in-depth gap rather than an open hole, and the proper fix — provisioning a dedicated least-privileged read-only role across migrations, docker, and helm — is its own change rather than an addition to this refactor.

File: SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+    'temp',
+    'apparentTemp',
+    'currentTemp',
+  ]);
+  const weatherTempCelsius =
+    rawWeatherTemp !== null ? ((rawWeatherTemp - 32) * 5) / 9 : null;
+
+  const rawWindSpeed = firstNum(weather, ['windSpeed']);
+  const weatherWindSpeedMps =
+    rawWindSpeed !== null ? rawWindSpeed * 0.44704 : null;
```
[the remark this answers is not in the document]
Comment (CodeWithCJ) [reply SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts:469]: Agreed, deferring the fix — tracked as #2000.

Garmin returns weather in the account's configured unit with no reliable marker in the payload, so the unconditional F→C and mph→m/s conversions do corrupt metric-account values. Fixing it properly means resolving the account measurement unit from the API; adding a magnitude heuristic instead would just be a guess. Left out of this PR rather than expanded into it. The `* 60` duration conversions are unchanged, as you noted.

**Comment by apedley** [comment]
Since we are pushing this to next release here are the bigger issues I found. I can do a full review later. I did this last night so line numbers might be different

### db

db_schema_backup.sql doesn't match the migrations. It looks like it was dumped from your dev db that had went through earlier versions of the migrations. The easiest fix for it is to merge main into dev, set up a clean db, run all the migrations and then regenerate the file.

Health Connect bone_mass is already in check_in_measurements.bone_mass_kg and needs to be migrated along with the garmini and withings data.

schemas/database/ExerciseEntries.zod.ts has exercise_id as nullable but it's NOT  NULL in db

HealthMetricSamples.zod.ts comment points at 20260801000000_consolidate_health_metric_samples.sql, the file is 20260730000000

idx_exercise_entry_laps_entry and idx_exercise_entry_hr_zones_entry duplicate their UNIQUE constraints, and idx_health_metric_samples_user_date / idx_daily_health_metrics_user_date are prefixes of theirs. All four can go 

### other

- FIT gps will enter all (0,0) points. extractGarminGPSPoints in garminTelemetryExtractors.ts:201 takes the metrics branch whenever they exist, but coords went into polyline. The polyline code is unreachable for FIT.
- garminHealthProcessor.ts:584 accepts item.hrvSummary or item.timestamp+item.rmssd but routes.py ssends hrvValue: [{time, data}] plus flat keys (last_night_avg, weekly_avg) so n o hrv row can ever be written and the per reading overnight data that was fetched gets discarded
- FIT laps never get saved since lapDto has no start time field and extractGarminLaps requires a startTime type field. LapMesg.startTime is right there
- On line 640, the function is expecting respirationValuesArray but it's not there. routes.py sends daily averages
- double check date calculations. I found at least one time zone boundary new Date problem in exerciseStatsService.ts:298
- lap timestamps are parsed in the wrong time zone. extractGarminLaps (garminTelemetryExtractors.ts:118) does new Date but this one is only an issue for people who have their server timezone set. still shold be fixed
- again with the best set comparisons.. we dont want to pair MAX(weight) with MAX(reps) because then you can easily get a set that never happened. Pick the single best row
- hasGpsTrack checks source === 'strava' but the pipeline stores 'Strava'

File: SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+      const idx = num(desc?.metricsIndex);
+      return idx !== null ? idx : -1;
+    };
+
+    const latIdx = getMetricIdx('directLatitude', 'latitude');
+    const lonIdx = getMetricIdx('directLongitude', 'longitude');
+    const altIdx = getMetricIdx('directElevation', 'elevation');
+    const hrIdx = getMetricIdx('directHeartRate', 'heartRate');
+    const speedIdx = getMetricIdx('directSpeed', 'speed');
+    const cadIdx = getMetricIdx('directRunCadence', 'cadence');
```
Comment (apedley) [inline SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts]: Same bug as the coordinates before. fitActivityTransform emits directDoubleCadence resulting in cadence being null on every point

File: SparkyFitnessGarmin/routes.py
Comment (apedley) [inline SparkyFitnessGarmin/routes.py:1286]: this throws away bp_entry's own timestamp and then it gets time stamped by garminHealthProcessor with T12:00:00Z. This means two measurements from the same day collapse

File: SparkyFitnessFrontend/src/pages/Reports/ExerciseReportsDashboard.tsx [hunk 12]
```diff
@@ -77,6 +124,20 @@ const ExerciseReportsDashboard = ({
...
 
+  const [statsInterval, setStatsInterval] = useState<
+    'day' | 'week' | 'month' | 'year'
+  >('day');
+
+  const { data: statsSummary } = useExerciseStatsSummary(
+    statsInterval,
+    startDate,
+    endDate,
+    undefined,
```
Comment (apedley) [inline SparkyFitnessFrontend/src/pages/Reports/ExerciseReportsDashboard.tsx]: should be activeUserId from useActiveUser()

File: SparkyFitnessServer/services/exerciseStatsService.ts [hunk 13]
```diff
@@ -0,0 +1,859 @@
...
+    );
+    const workoutCount = parseInt(String(totalsRow.workout_count || '0'), 10);
+    const totalElevationGainMeters = Math.round(
+      parseFloat(String(totalsRow.total_elevation_gain_meters || '0'))
+    );
+    const avgHeartRate = totalsRow.avg_heart_rate
+      ? Math.round(parseFloat(String(totalsRow.avg_heart_rate)))
+      : null;
+
+    const strengthSql = `
```
Comment (apedley) [inline SparkyFitnessServer/services/exerciseStatsService.ts:158]: Missing AND LOWER(category) = LOWER($4)

File: shared/src/schemas/api/ExerciseStats.api.zod.ts [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+  distanceMaxMeters: z.number().optional(),
+  distanceStandard: z
+    .enum(["1k", "1mi", "5k", "10k", "15k", "half_marathon", "marathon", "custom"])
+    .optional(),
+  startDate: z.string().optional(),
+  endDate: z.string().optional(),
+  minPaceSecondsPerKm: z.number().optional(),
+  maxPaceSecondsPerKm: z.number().optional(),
+  minAvgHeartRate: z.number().optional(),
+  maxAvgHeartRate: z.number().optional(),
```
Comment (apedley) [inline shared/src/schemas/api/ExerciseStats.api.zod.ts]: These 4 minPaceSecondsPerKm, maxPaceSecondsPerKm, minAvgHeartRate and maxAvgHeartRate aren't ever used i believe

File: SparkyFitnessFrontend/src/pages/Diary/EditExerciseEntryDialog.tsx [hunk 15]
```diff
@@ -119,29 +322,266 @@ const EditExerciseEntryDialog = ({
...
+    if (extracted.avgMovingSpeed !== '')
+      setAvgMovingSpeedInput(extracted.avgMovingSpeed);
+    if (extracted.weatherTemp !== '')
+      setWeatherTempInput(extracted.weatherTemp);
+    if (extracted.windSpeed !== '') setWindSpeedInput(extracted.windSpeed);
+    if (extracted.humidity !== '') setHumidityInput(extracted.humidity);
+    if (entry.duration_minutes != null && entry.duration_minutes > 0) {
+      setDurationInput(Math.round(Number(entry.duration_minutes) * 100) / 100);
+    }
+  }, [open, entry, activityData]);
```
Comment (apedley) [inline SparkyFitnessFrontend/src/pages/Diary/EditExerciseEntryDialog.tsx]: any refresh of activityData will overwrite what you write

File: SparkyFitnessServer/services/garmin/garminHealthProcessor.ts [hunk 16]
```diff
@@ -0,0 +1,1243 @@
...
+          chronic_training_load: chronicLoad,
+          acwr_ratio: acwrRatio,
+        });
+        processedResults.push({
+          type: 'daily_health_metrics',
+          status: 'success',
+          date: dateStr,
+        });
+      } catch (err) {
+        log(
```
Comment (apedley) [inline SparkyFitnessServer/services/garmin/garminHealthProcessor.ts:490]: Its good to log but this doesnt push the error into processedResults so its marked as success

File: SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+      zone_lower_bpm: firstNum(zone, [
+        'zone_lower_bpm',
+        'zoneLowBoundary',
+        'zone_low_boundary',
+        'lowerBound',
+      ]),
+      seconds_in_zone:
+        firstInt(zone, ['seconds_in_zone', 'secsInZone', 'secs_in_zone']) ?? 0,
+    }))
+    .filter((zone) => zone.zone_index > 0)
```
Comment (apedley) [inline SparkyFitnessServer/services/garmin/garminTelemetryExtractors.ts]: this drops any zone (if its zero indexed which it probably is)

**Review by apedley — COMMENTED** [review]
Could really use tests over the hrv and respiration parts of garminHealthProcessor. I think it would catch a lot of this. If you're interested I can come up with one and PR it to this branch

**Comment by CodeWithCJ** [comment]
@apedley Please review and feel free to submit changes directly to this branch if any.
